Retry HTTP for status >= 500 (exponential backoff)

attilapiros opened the pull request:
## Description

Fix #3087 
This PR introduces configurable retry with exponential backoff for HTTP operations.
As according to the [API conventions](https://github.com/fabric8io/kubernetes-client/issues/3087) when the HTTP status code is one of the followings:

- 500 StatusInternalServerError
- 503 StatusServiceUnavailable
- 504 StatusServerTimeout

Then the suggested client recovery behavior is "retry with exponential backoff".

Although in case of 504 it is "Increase the value of the timeout param and retry with exponential backoff" but in this PR I would only focus on the retries. That can be implemented in a follow up PR.

For the intention behind please check: https://github.com/fabric8io/kubernetes-client/issues/3087.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

shawkins commented:
I think the code looks good, a general question would be around the existing backoff strategy in https://github.com/fabric8io/kubernetes-client/blob/df5d80235386e339f6ac1a3fbd8ecb2bad2cfa9c/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/AbstractWatchManager.java#L127 

Should there be consistency between the properties / strategy?  Watch only uses interval and limit.  There's also a hardcoded value of 5 limiting the exponential scaling to 32x.

attilapiros commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/ExponentialBackoffIntervalCalculator.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
```
Should I update the year? 2015 -> 2021?

shawkins replied:
The license header check I believe expects 2015.

attilapiros commented:
I have renamed the properties to follow the existing naming used at watchers:
- kubernetes.request.retry.backoffLimit`
- kubernetes.request.retry.backoffInterval`

And extracted the interval calculation into `ExponentialBackoffIntervalCalculator` which used by both `OperationSupport` and `AbstractWatchManager`.

This new class is not tested separately as the functionality is already tested by the existing watcher's unit test and it's logic is lightweight:
`https://github.com/fabric8io/kubernetes-client/blob/3c95a6f7da2271392a76ae806d475018d33b66d3/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/AbstractWatchManagerTest.java#L83`
But we can revisit this decision if you feel like.

The force-push was done because of the rebase on the top of `master`.

attilapiros commented:
I think the integration test failure is unrelated, probably a flaky test:
```
Error:  Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.729 s <<< FAILURE! - in io.fabric8.kubernetes.PersistentVolumeClaimIT
Error:  io.fabric8.kubernetes.PersistentVolumeClaimIT.update  Time elapsed: 0.182 s  <<< ERROR!
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at:
 https://10.1.1.35:8443/api/v1/namespaces/itest-a2b75c94/persistentvolumeclaims/persistentvolumeclaims-update. 
Message: the server rejected our request due to an error in our request. 
Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[], group=null, kind=null, name=null,
 retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=the server rejected our request due 
 to an error in our request, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null,
  additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.PersistentVolumeClaimIT.update(PersistentVolumeClaimIT.java:66)
```

attilapiros commented:
I looked around and found this:
https://github.com/fabric8io/kubernetes-client/pull/3136#issuecomment-843243513

So this error was a known problem before which tried to resolved by https://github.com/fabric8io/kubernetes-client/pull/3148

My changes are on the top of the current HEAD so I have these commit:
https://github.com/attilapiros/kubernetes-client/commit/d1ab437ec86e809bdcb3a45f8dcae6e3583413e3
https://github.com/attilapiros/kubernetes-client/commit/390de033258121038d20676db5b669b5a878aa35

cc @shawkins @manusa could you please take a look to this?

attilapiros commented:
Big thanks for the rerun to anybody who helped me! :)

shawkins commented:
> could you please take a look to this?

Yes that is the same issue an edit is just a different from of patch(item).  That test should be updated to remove the possibility of a 422.

shawkins reviewed (approved):
LGTM

manusa commented on README.md:
```diff
@@ -92,6 +92,8 @@ System properties are preferred over environment variables. The following system
 | `kubernetes.watch.reconnectLimit` / `KUBERNETES_WATCH_RECONNECTLIMIT` | Number of reconnect attempts (-1 for infinite) | `-1` |
 | `kubernetes.connection.timeout` / `KUBERNETES_CONNECTION_TIMEOUT` | Connection timeout in ms (0 for no timeout) | `10000` |
 | `kubernetes.request.timeout` / `KUBERNETES_REQUEST_TIMEOUT` | Read timeout in ms | `10000` |
+| `kubernetes.request.retry.backoffLimit` / `KUBERNETES_REQUEST_RETRY_BACKOFFLIMIT_SYSTEM_PROPERTY` | Number of retry attempts | `0` |
+| `kubernetes.request.retry.backoffInterval` / `KUBERNETES_REQUEST_RETRY_BACKOFFINTERVAL_SYSTEM_PROPERTY` | Retry initial backoff interval in ms | `1000` |
```
```suggestion
| `kubernetes.request.retry.backoffLimit` / `KUBERNETES_REQUEST_RETRY_BACKOFFLIMIT` | Number of retry attempts | `0` |
| `kubernetes.request.retry.backoffInterval` / `KUBERNETES_REQUEST_RETRY_BACKOFFINTERVAL` | Retry initial backoff interval in ms | `1000` |
```

attilapiros replied:
Thanks!

manusa reviewed (approved):
LGTM, thx!

geoand commented:
Cool feature!

I have a question, do these new properties come into play when a `java.net.ConnectException: Connection refused (Connection refused)` is thrown by the client?
Looking at the code, it doesn't look like it. Would it not make sense to handle this case as well?

rohanKanojia commented:
@attilapiros @shawkins @manusa : Could you please share your thoughts on this?

shawkins commented:
> Looking at the code, it doesn't look like it

Yes, it will just throw the IOExeption.

> Would it not make sense to handle this case as well?

Yes, that's an assumption made in other places, like with informers https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L416

FWIW the code in fabric8 is a little different for watches/informers wrt connection refused - if the initial watch/list (which would be subject to the common retry logic) succeeds, then any later IOExceptions will be assumed to be recoverable - indefinitely in the case of informers.

geoand commented:
Thanks for the information.

Should I open a new issue for this feature request?

shawkins commented:
> Should I open a new issue for this feature request?

Yes please.

attilapiros commented:
Sorry I am away from my computer but next week latest I can check it

geoand commented:
https://github.com/fabric8io/kubernetes-client/issues/3291 is the new issue
